builtin/providers/ibmcloud/resource_ibmcloud_infra_virtual_guest.go

```diff
@@ -0,0 +1,1061 @@
+			},
+		)
+	}
+
+	// Configure secondary IPs
+	secondaryIPCount := d.Get("secondary_ip_count").(int)
+	if secondaryIPCount > 0 {
+		if privateNetworkOnly {
+			return fmt.Errorf("Unable to configure public secondary addresses with a private_network_only option")
+		}
+		staticIPItems, err := services.GetProductPackageService(sess).
+			Id(*template.PackageId).
+			Mask("id,capacity,description,units,keyName,prices[id,categories[id,name,categoryCode]]").
+			Filter(filter.Build(filter.Path("items.keyName").Eq(strconv.Itoa(secondaryIPCount) + "_PUBLIC_IP_ADDRESSES"))).
+			GetItems()
+		if err != nil {
+			return fmt.Errorf("Error generating order template: %s", err)
+		}
+		if len(staticIPItems) == 0 {
+			return fmt.Errorf("No product items matching %d_PUBLIC_IP_ADDRESSES could be found", secondaryIPCount)
+		}
+
+		template.Prices = append(template.Prices,
+			datatypes.Product_Item_Price{
+				Id: staticIPItems[0].Prices[0].Id,
+			},
+		)
+	}
+
+	// GenerateOrderTemplate omits UserData, subnet, and maxSpeed, so configure virtual_guest.
+	template.VirtualGuests[0] = opts
+
+	order := &datatypes.Container_Product_Order_Virtual_Guest{
+		Container_Product_Order_Hardware_Server: datatypes.Container_Product_Order_Hardware_Server{Container_Product_Order: template},
+	}
+
+	orderService := services.GetProductOrderService(sess)
+	receipt, err := orderService.PlaceOrder(order, sl.Bool(false))
+	if err != nil {
+		return fmt.Errorf("Error ordering virtual guest: %s", err)
+	}
+	id = *receipt.OrderDetails.VirtualGuests[0].Id
+
+	d.SetId(fmt.Sprintf("%d", id))
+
+	log.Printf("[INFO] Virtual Machine ID: %s", d.Id())
+
+	// Set tags
+	err = setGuestTags(id, d, meta)
+	if err != nil {
+		return err
+	}
+
+	// wait for machine availability
+
+	_, err = WaitForVirtualGuestAvailable(d, meta)
+
+	if err != nil {
+		return fmt.Errorf(
+			"Error waiting for virtual machine (%s) to become ready: %s", d.Id(), err)
+	}
+
+	return resourceIBMCloudInfraVirtualGuestRead(d, meta)
+}
+
+func resourceIBMCloudInfraVirtualGuestRead(d *schema.ResourceData, meta interface{}) error {
+	service := services.GetVirtualGuestService(meta.(ClientSession).SoftLayerSession())
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid ID, must be an integer: %s", err)
+	}
+
+	result, err := service.Id(id).Mask(
+		"hostname,domain,startCpus,maxMemory,dedicatedAccountHostOnlyFlag," +
+			"primaryIpAddress,primaryBackendIpAddress,privateNetworkOnlyFlag," +
+			"hourlyBillingFlag,localDiskFlag," +
+			"userData[value],tagReferences[id,tag[name]]," +
+			"datacenter[id,name,longName]," +
+			"primaryNetworkComponent[networkVlan[id]," +
+			"primaryVersion6IpAddressRecord[subnet,guestNetworkComponentBinding[ipAddressId]]," +
+			"primaryIpAddressRecord[subnet,guestNetworkComponentBinding[ipAddressId]]]," +
+			"primaryBackendNetworkComponent[networkVlan[id]," +
+			"primaryIpAddressRecord[subnet,guestNetworkComponentBinding[ipAddressId]]]",
+	).GetObject()
```

Comment:
Although similar one we saw for the ssh key in the remarks above and the answer holds true here as well. Still worth the explanation as in this case, in addition to bringing less data, we also need to bring all required data with just single API call. We are able to pull in [_Relational_](http://sldn.softlayer.com/reference/datatypes/SoftLayer_Virtual_Guest) properties of Virtual Guest  also in the same API call which will not be possible without masking and getting to multiple API calls. Again this is supported at the SoftLayer API level and hence we are using that. I think [SDK](http://github.com/softlayer/softlayer-go) provides this high level API and as there can be so many combinations of fields that a user may want, leaves it to the client to ask what it needs.